src/ldapauth.c

```diff
@@ -0,0 +1,772 @@
+
+	while (next_free_slot == ldap_first_taken_slot) {
+		if (ldap_poll() == 0) {
+			/* Sleep a bit between consequent queue checks to avoid consuming too much CPU */
+			usleep(LDAP_QUEUE_WAIT_SLEEP_MCS);
+		}
+	}
+
+	pthread_mutex_lock(&ldap_queue_tail_mutex);
+
+	request = &ldap_auth_queue[ldap_first_free_slot];
+
+	request->client = client;
+	request->connect_time = client->connect_time;
+	request->status = LDAP_STATUS_IN_PROGRESS;
+	memcpy(&request->remote_addr, &client->remote_addr, sizeof(client->remote_addr));
+	safe_strcpy(request->username, client->login_user->name, MAX_USERNAME);
+	safe_strcpy(request->password, passwd, MAX_PASSWORD);
+	/* Reset value of ldap parameters */
+	free_ldap_parameters(request);
+
+	ldap_first_free_slot = next_free_slot;
+
+	pthread_cond_signal(&ldap_data_available);
+	pthread_mutex_unlock(&ldap_queue_tail_mutex);
+}
+
+/*
+ * Checks for completed auth requests, returns amount of requests handled.
+ * The function is called only from the main thread.
+ */
+int ldap_poll(void)
+{
+	struct ldap_auth_request *request;
+	int count = 0;
+
+	while (ldap_first_taken_slot != ldap_first_free_slot) {
+		request = &ldap_auth_queue[ldap_first_taken_slot];
+
+		if (request->status == LDAP_STATUS_IN_PROGRESS) {
+			/* When still-in-progress slot is found there is no need to continue
+			 * the loop since all further requests will be in progress too.
+			 */
+			break;
+		}
+
+		if (is_valid_socket(request)) {
+			ldap_auth_finish(request);
+		}
+
+		count++;
+		ldap_first_taken_slot = (ldap_first_taken_slot + 1) % LDAP_REQUEST_QUEUE_SIZE;
+	}
+
+	return count;
+}
+
+
+/*
+ * The authentication thread function.
+ * Performs scanning the queue for new requests and calling LDAP for them.
+ */
+static void *ldap_auth_worker(void *arg)
+{
+	int current_slot = ldap_first_taken_slot;
+	struct ldap_auth_request *request;
+
+	while (true) {
+		/* Wait for new data in the queue */
+		pthread_mutex_lock(&ldap_queue_tail_mutex);
+
+		while (current_slot == ldap_first_free_slot) {
+			pthread_cond_wait(&ldap_data_available, &ldap_queue_tail_mutex);
+		}
+
+		pthread_mutex_unlock(&ldap_queue_tail_mutex);
+
+		log_debug("ldap_auth_worker(): processing slot %d", current_slot);
+
+		/* We have at least one request in the queue */
+		request = &ldap_auth_queue[current_slot];
+		current_slot = (current_slot + 1) % LDAP_REQUEST_QUEUE_SIZE;
+
+		/* If the socket is already in the wrong state or reused then ignore it.
+		 * This check is not safe and should not be trusted (the socket state
+		 * might change exactly after it), but it helps to quickly filter out invalid
+		 * sockets and thus save some time.
+		 */
+		if (!is_valid_socket(request)) {
+			log_debug("ldap_auth_worker(): invalid socket in slot %d", current_slot);
+			request->status = LDAP_STATUS_FAILED;
+			continue;
+		}
+
+		if (check_ldap_auth(request)) {
+			request->status = LDAP_STATUS_SUCCESS;
+		} else {
+			request->status = LDAP_STATUS_FAILED;
+		}
```

Comment:
It seems you copied this file mostly from the PAM logic. But as I found out a few days ago, this logic is really not thread safe afaict: https://github.com/pgbouncer/pgbouncer/issues/1054#issuecomment-2083095520

Reply:
Actually this `is_valid_socket` call is also completely unsynchronized with the main thread, so that also returns garbage.

Reply:
I think the worker thread should never read directly from `request->client` and instead the main thread should mark a request as cancelled by setting some flag on the request (while having a lock) .

Reply:
Yes, the reading and writing of `request->status` and `is_valid_socket(request)` are not synchronized.
I guess the reason why they still work is that the readings of structure `request->status` are only from main thread, and writings are from the `ldap_auth_worker()`. And the writings of `client->state` and `client->connect_time` are from main thread, and `ldap_auth_worker()` only does the readings.
The `request->status` is an internal structure of ldap or pam, so I think it is easy to add lock for it.
But for `is_valid_socket`, it is not easy to add.

Reply:
I reconsidered the access of `request->status` and `client->state` or `client->connect_time`.
I think there is no need a lock on `request->status` because the only writing of this structure `request->status` which is type of int, is the thread worker, and the main thread only check the value change, and if the value changes either to `LDAP_STATUS_SUCCESS` or `LDAP_STATUS_FAILED`, then the main thread will continue to do something, and the main thread only cares about the changing of the value `request->status`. The sequence is not important, because if main got a unchanged value, it will continue to check on the next 100ms. If it got a changed value, that means the worker thread truly changed the value.
And this works the same for the checking function  `is_valid_socket()` which will access the value of `client->state` and `client->connect_time`which is type int and int64. The worker thread only cares about the changing, (the main thread will not change the value at the same time in the same main thread, so we don't need to care about main thread accessing)but the sequence is not important, and if it changes, the `is_valid_socket()` will be false.
The basis of this is that the writing back into structure `int` and `int64` will be done in one instruction. If we are trying to check some more complicated structure which could not be write back in 1 instruction, then we definitely need a lock.

Reply:
> The basis of this is that the writing back into structure int and int64 will be done in one instruction. 

It doesn't matter if it is one instruction or not. For writes to any value to be correctly synced to another thread you need to write it using an atomic operation, or use a lock to synchronize it. If you do otherwise, that can cause undefined behaviour due to data races. Hardware wise, the reason for this is that the two threads might be running on different cores. These cores have often have separate L1 caches. So if one core changes the value, the other core might still read the previous value from its cache. Writing to the variable using an atomic operation (or using a lock) will make sure the other cores their caches get updated/flushed when such a write happens.

Reply:
The comments before calling function `is_valid_socket()` in the worker thread explains the reason:
```
/* If the socket is already in the wrong state or reused then ignore it.
 * This check is not safe and should not be trusted (the socket state
 * might change exactly after it), but it helps to quickly filter out invalid
 * sockets and thus save some time.
 */
```
I don't know if you agree with it, even if it is not safe, but it indicates that the state has changed after it really changed. If you don't agree with it, then I think I can remove the check, because it should not impact the final result. Because when `client->state` has changed into `CL_LOGIN`, then it will not try to receive data from socket, then only when timeout, the client will be disconnected. Then nothing happens when we call `sbuf_continue(&client->sbuf);` after `disconnect_client` is called.

Reply:
It definitely seems a lot better to remove it than keep it. Your newly pushed code looks safe from a race condition perspective.